src/lib.rs

```diff
@@ -99,21 +85,76 @@ fn exception_handler(context: &arch::TrapFrame) -> ! {
     );
     println!("{:x?}", context);
 
-    let backtrace = crate::arch::backtrace_internal(context.s0 as u32, 0);
-    if backtrace.iter().filter(|e| e.is_some()).count() == 0 {
-        println!("No backtrace available - make sure to force frame-pointers. (see https://crates.io/crates/esp-backtrace)");
+    let backtrace = Backtrace::new(core::ptr::addr_of!(context) as u32, 0);
+    print_backtrace(backtrace);
+
+    halt();
+}
+
+#[derive(Clone, Copy)]
+pub struct Backtrace {
+    /// Frame pointer
+    fp: u32,
+
+    /// Depth counter to avoid looping on cyclic recursion.
+    depth: usize,
+}
+
+impl Backtrace {
+    #[inline(always)]
+    fn new(fp: u32, suppress: usize) -> Self {
+        let mut this = Backtrace { fp, depth: 0 };
+
+        // We need to skip the first few frames here, otherwise returning to the caller might cause
+        // problems.
+        for _ in 0..suppress {
```

Comment:
This is where the magic happens. Looks like returning from `backtrace` pops the first stack frame we want to read 😅